### ISSUE TITLE
What is the correct way to obtain the last step infomation when skip is applied to controls?

### ISSUE THREAD

**Issue by liuyxpp** [JuliaAI/IterationControl.jl#38]
Example below:

```julia
nskip = 100
dispaly_control = IterationControl.skip(Info(m->"F: $(F(m))"), predicate=nskip)
number_control = IterationControl.skip(WithNumberDo(), predicate=nskip)
loss_control = IterationControl.skip(WithLossDo(), predicate=nskip)
IterationControl.train!(model,
                            Step(1),
                            number_control,
                            dispaly_control,
                            loss_control,
                            NumberLimit(config.scft.max_iter),
                            InvalidValue(),
                            TimeLimit(0.5),
                            Threshold(config.scft.tol),
                            NumberSinceBest(500),
                            Patience(500),
                            verbosity=1)
```

A typical output is

```julia
[ Info: number: 1
[ Info: F: 4.967735136304276
[ Info: loss: 0.5752192428897245
[ Info: number: 2
[ Info: F: 4.004661017907266
[ Info: loss: 0.6971497109719819
[ Info: number: 3
[ Info: F: 3.9854332008041666
[ Info: loss: 0.02397907003906548
[ Info: number: 4
[ Info: F: 3.985139894743944
[ Info: loss: 0.0007433714037504332
[ Info: number: 5
[ Info: F: 3.98513319008801
[ Info: loss: 2.111282679787922e-5
[ Info: Stop triggered by EarlyStopping.Threshold(1.0e-5) stopping criterion.
```

Two questions: 

1. Why number is displayed as 1,2... but not 1*nskip, 2*nskip, ...
2. I'd also like to print the same information for the last step (i.e. when the stoping criteria is triggered). How to achieve that?

Thanks!

**Comment by ablaom** [comment]
Very keen observation. And thank you for your query!

So "number" is not in general the total number of control cycles if there is skipping. Rather it is the number of times the `WithNumberDo` control has been applied. When you skip, the control is not applied and so its state is not updated. This behaviour is consistent with the detailed docstring:

---

    WithNumberDo(f=n->@info("number: $n"), stop_if_true=false, stop_message=nothing)

An iteration control, as in, `WithNumberDo(n->put!(my_channel, n))`.

Call `f(n)`, where `n` is one more than the number of previous applications of the
  control (so, `n = 1, 2, 3,` ...).

  If `stop_if_true` is true, then trigger an early stop if the value returned by `f`
  is `true`, logging the stop_message if specified.

---

So "number" is a variable "local" to the control, not a global count of control cycle applications. 

However, I agree, this is not the behaviour many would expect. And, if we agree this is also not the most *useful* behaviour, then I would be happy to look into a change. What do you think?


Regarding the second question: The issue here is that your printing control is not executed every cycle; if a stop is triggered in a cycle with no printing, then printing does not happen. This much I guess you understand already. 

The final "state" of each control is in the return value of `IterationControl.train!`. However this "state" does not always contain the information you want (eg, loss in the case of `WithLossDo`) - just the minimum needed to make the control work. We could easily add this information where extra allocation is unlikely to be a concern. 

A workaround for now is to include a `Callback` control to write the information you want to some container, for later access at the end of training, as in the  readme example. 

Does that help?

**Comment by liuyxpp** [comment]
For the first question,  I think we should make it clear somewhere in the doc that two numbers, i.e. the global iteration number counter and the internal control number. Even better, if we can figure out a way to broadcast the global iteration number to all controls because this number is so important?

Also, the termination state of an iteration is very important. Do you think it is well deserved to make a special control to catch it? At present, I will implement a `Callback` to do the job.

Thanks for your help!

**Comment by ablaom** [comment]
Regarding issues raised in this thread:

**Displaying/accesssing the final loss.** In 0.3.2 the final loss (respectively, training loss) is logged to `Info` at the default `verbosity=1`. These losses are also added to the "report" of any `WithLossDo` (respectively `WithTrainingLossDo`) control. Reports are explained [here](https://github.com/JuliaAI/IterationControl.jl#verbose-logging-and-inspecting-control-reports).

In the following example the final loss is logged to `Info`, even though the `WithLossDo()` control wrapped in `skip` never sees this particular loss:

```julia
model = IterationControl.SquareRooter(4)

julia> report =  IC.train!(model,
       Step(1),
       Threshold(0.0001),
       WithNumberDo(),
       IterationControl.skip(WithLossDo(), predicate=3));
[ Info: number: 1
[ Info: number: 2
[ Info: number: 3
[ Info: loss: 0.002439396192741583
[ Info: number: 4
[ Info: final loss: 3.716891878724482e-7
[ Info: final training loss: 0.0006096631752661175
[ Info: Stop triggered by Threshold(0.0001) stopping criterion. 
```

To *access* the final loss (not just display it), you can add a second silent `WithLossDo` control and inspect it's report:

```julia
model = IterationControl.SquareRooter(4)
reports =  IC.train!(model,
       Step(1),
       Threshold(0.0001),
       WithNumberDo(),
       IterationControl.skip(WithLossDo(), predicate=3),
       WithLossDo(_->nothing));

julia> last(reports[5])
(loss = 3.716891878724482e-7, done = false, log = "")
```

**The meaning of "number"** Addressed in [this breaking PR](https://github.com/JuliaAI/IterationControl.jl/pull/44). In future "number" is the number of completed control cycles.
